src/router/slang/slsh/doc/man/slsh.1
.\" This manpage has been automatically generated by docbook2man 
.\" from a DocBook document.  This tool can be found at:
.\" <http://shell.ipoline.com/~elmert/comp/docbook2X/> 
.\" Please send any bug reports, improvements, comments, patches, 
.\" etc. to Steve Cheng <steve@ggi-project.org>.
.TH "SLSH" "1" "28 March 2011" "" ""

.SH NAME
slsh \- Interpreter for S-Lang scripts
.SH SYNOPSIS

\fBslsh\fR [ \fB--help\fR ] [ \fB--version\fR ] [ \fB-g\fR ] [ \fB-n\fR ] [ \fB--init \fIfile\fB\fR ] [ \fB--no-readline\fR ] [ \fB-e \fIstring\fB\fR ] [ \fB-i\fR ] [ \fB-q, --quiet\fR ] [ \fB-t\fR ] [ \fB-v\fR ] [ \fB-|\fIscript-file args...\fB\fR ]

.SH "DESCRIPTION"
.PP
\fBslsh\fR is a simple program for interpreting S-Lang scripts.  It
supports dynamic loading of S-Lang modules and includes a readline
interface for interactive use.
.SH "OPTIONS"
.TP
\fB--help\fR
Show a summary of options
.TP
\fB--version\fR
Show \fBslsh\fR version information
.TP
\fB-g\fR
Compile with debugging code, tracebacks, etc
.TP
\fB-n\fR
Don't load the personal initialization file
.TP
\fB--init \fIfile\fB\fR
Use this file instead of ~/.slshrc
.TP
\fB--no-readline\fR
Do not use a readline interface for the interactive mode
.TP
\fB-e \fIstring\fB\fR
Execute ``string'' as S-Lang code.
.TP
\fB-i\fR
Force interactive mode.  Normally \fBslsh\fR will go into
interactive mode if both stdin and stdout are attached to a
terminal.
.TP
\fB-q, --quiet\fR
Startup quietly by not printing the version and copyright
information.
.TP
\fB-t\fR
Normally, \fBslsh\fR will call slsh_main if it is defined.  This
option prevents that from happening making it useful for
checking for syntax error.
.TP
\fB-v\fR
Show verbose loading messages.  This is useful for seeing what
files are being loaded.
.SH "INITIALIZATION"
.PP
Upon startup, the program will try to load \fIslsh.rc\fR as
follows. If either SLSH_CONF_DIR or
SLSH_LIB_DIR environment variables exist, then \fBslsh\fR will
look look in the corresponding directories for \fIslsh.rc\fR\&.
Otherwise it will look in:

\fI$(prefix)/etc/\fR   (as specified in the Makefile)

\fI/usr/local/etc/\fR

\fI/usr/local/etc/slsh/\fR

\fI/etc/\fR

\fI/etc/slsh/\fR
.PP
The \fIslsh.rc\fR file may load other files from slsh's library
directory in the manner described below.
.PP
Once \fIslsh.rc\fR has been loaded, \fBslsh\fR will load
$HOME/.slshrc if present.  Finally, it will load the
script specified on the command line.  If the name of the script is
-, then it will be read from stdin.  If the script name is
not present, or a string to execute was not specified using the -e
option, then \fBslsh\fR will go into interactive mode and read input from
the terminal.  If the script is present and defines a function
called slsh_main, that function will be called.
.SH "LOADING FILES"
.PP
When a script loads a file via the built-in evalfile function
or the require function (autoloaded by slsh.rc), the file is
searched for along the SLSH_PATH as specified in the Makefile.  An
alternate path may be specified by the SLSH_PATH environment
variable.
.PP
The search path may be queried and set during run time via the
get_slang_load_path and set_slang_load_path functions, e.g.,

.nf
   set_slang_load_path ("/home/bill/lib/slsh:/usr/share/slsh");
.fi
.SH "INTERACTIVE MODE"
.PP
When \fBslsh\fR is invoked without a script or is given the \fB-i\fR
command line argument, it will go into into interactive mode.  In
this mode, the user will be prompted for input.  The program will
leave this mode and exit if it sees an EOF (Ctrl-D) or the user
exits by issuing the quit command.
.PP
If an uncaught exception occurs during execution of a command, the
error message will be shown and the user will be prompted for more
input.
.PP
Any objects left on the stack after a command will be printed and
the stack cleared.  This makes interactive mode useful as a
calculator, e.g.,

.nf
     slsh> 3*10;
     30
     slsh> x = [1:20];
     slsh> sum (sin(x)-cos(x));
     0.458613
     slsh> quit;
.fi
Note that in this mode, variables are automatically declared.
.PP
The interactive mode also supports command logging.  Logging is
enabled by the start_log function.  The stop_log
function will turn off logging.  The default file where logging
information will be written is \fIslsh.log\fR\&.  An alternative
may be specified as an optional argument to the start_log
function:

.nf
     slsh> start_log;
     Logging input to slsh.log
\&.
\&.
     slsh> stop_log;
     slsh> start_log("foo.log");
     Logging input to foo.log
\&.
\&.
     slsh> stop_log;
     slsh> start_log;
     Logging input to foo.log
.fi
.PP
Similarly, the save_input function may be used to save the
previous input to a specified file:

.nf
     slsh> save_input;
     Input saved to slsh.log
     slsh> save_input ("foo.log");
     Input saved to foo.log
.fi
.PP
As the above examples indicate, lines must end in a semicolon.  This
is a basic feature of the language and permits commands to span
multiple lines, e.g.,

.nf
     slsh> x = [
            1,2,3,
            4,5,6];
     slsh> sum(x);
.fi
For convenience some users prefer that commands be automatically
terminated with a semicolon.  To have a semicolon silently appended
to the end of an input line, put the following in
\fI$HOME/.slshrc\fR file:

.nf
    #ifdef __INTERACTIVE__
    slsh_append_semicolon (1);
    #endif
.fi
.PP
The interactive mode also supports shell escapes.  To pass a command
to the shell, prefix it with !, e.g.,

.nf
    slsh> !pwd
    /grandpa/d1/src/slang2/slsh
    slsh> !cd doc/tm
    slsh> !pwd
    /grandpa/d1/src/slang2/slsh/doc/tm
.fi
.PP
Finally, the interactive mode supports a help and
apropos function:

.nf
    slsh> apropos list
    apropos list ==>
    List_Type
    list_append
    list_delete
\&.
\&.
    slsh> help list_append
    list_append

     SYNOPSIS
       Append an object to a list

     USAGE
       list_append (List_Type, object, Int_Type nth)
\&.
\&.
.fi
For convenience, the help and apropos functions
do not require the syntactic constraints of the other functions.
.SH "READLINE HISTORY MECHANISM"
.PP
By default, \fBslsh\fR is built to use the S-Lang readline interface,
which includes a customizable command completion and a history mechanism.
When \fBslsh\fR (or any S-Lang application that makes use of this
feature) starts in interactive mode, it will look for a file in the
user's home directory called \fI\&.slrlinerc\fR and load it if
present.  This file allows the user to customize the readline
interface and enable the history to be saved between sessions.  As
an example, here is a version of the author's
\fI\&.slrlinerc\fR file:

.nf
     % Load some basic functions that implement the history mechanism
     () = evalfile ("rline/slrline.rc");
     % The name of the history file -- expands to .slsh_hist for slsh
     RLine_History_File = "$HOME/.${name}_hist";

     % Some addition keybindings.  Some of these functions are defined
     % in rline/editfuns.sl, loaded by rline/slrline.rc
     rline_unsetkey ("^K");
     rline_setkey ("bol",   "^B");
     rline_setkey ("eol",   "^E");
     rline_setkey (&rline_kill_eol,  "^L");
     rline_setkey (&rline_set_mark,  "^K^B");
     rline_setkey (&rline_copy_region, "^Kk");
     rline_setkey (&rline_kill_region, "^K^V");
     rline_setkey (&rline_yank,  "^K^P");
     rline_setkey ("redraw",   "^R");

     #ifexists rline_up_hist_search
     % Map the up/down arrow to the history search mechanism
     rline_setkey (&rline_up_hist_search, "\\e[A");
     rline_setkey (&rline_down_hist_search, "\\e[B");
     #endif

     #ifexists rline_edit_history
     rline_setkey (&rline_edit_history, "^Kj");
     #endif

     % Add a new function
     private define double_line ()
     {
        variable p = rline_get_point ();
        variable line = rline_get_line ();
        rline_eol ();
        variable pend = rline_get_point ();
        rline_ins (line);
        rline_set_point (pend + p);
     }
    rline_setkey (&double_line,  "^K^L");
.fi
.SH "MISCELLANEOUS SCRIPTS"
.PP
Several useful example scripts are located in
\fI$prefix/share/slsh/scripts/\fR, where $prefix represents the
\fBslsh\fR installation prefix (\fI/usr\fR,
\fI/usr/local\fR,...).  These scripts include:
.TP
\fBsldb\fR
A script that runs the S-Lang debugger. 
.TP
\fBjpegsize\fR
Reports the size of a jpeg file. 
.TP
\fBsvnsh\fR
A shell for browsing an SVN repository. 
.SH "AUTHOR"
.PP
The principal author of \fBslsh\fR is John E. Davis <www.jedsoft.org>\&.
The interactive mode was provided by Mike Noble.  The S-Lang library
upon which \fBslsh\fR is based is primarily the work of John E. Davis
with help from many others.
.PP
This manual page was originally written by Rafael Laboissiere for
the Debian system (but may be used by others).
.PP
Permission is granted to copy, distribute and/or modify
this document under the terms of the GNU General Public License,
Version 2 any later version published by the Free Software
Foundation.
.PP
On Debian systems, the complete text of the GNU General Public
License can be found in \fI/usr/share/common-licenses/GPL\fR


src/router/slang/slsh/slsh.c
/*
Copyright (C) 2005-2011 John E. Davis

This file is part of the S-Lang Library.

The S-Lang Library is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License as
published by the Free Software Foundation; either version 2 of the
License, or (at your option) any later version.

The S-Lang Library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License
along with this library; if not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307,
USA.
*/

#include "config.h"
#include <stdio.h>
#include <stdlib.h>
#ifdef __WIN32__
# include <windows.h>
# include <io.h>
#endif

#include <sys/stat.h>

#ifdef HAVE_UNISTD_H
# include <unistd.h>
#endif
#include <string.h>
#include <signal.h>
#include <slang.h>

static char *Slsh_Version = "0.8.6-0";
#define SLSHRC_FILE "slsh.rc"
#include "slsh.h"

#ifndef SLSH_CONF_DIR_ENV
# define SLSH_CONF_DIR_ENV "SLSH_CONF_DIR"
#endif

#ifndef SLSH_PATH_ENV
# define SLSH_PATH_ENV "SLSH_PATH"
#endif

#ifndef SLSH_LIB_DIR_ENV
# define SLSH_LIB_DIR_ENV "SLSH_LIB_DIR"
#endif

#if defined(REAL_UNIX_SYSTEM) || defined(__APPLE__)
/* # define DEFAULT_LIBRARY_PATH "/usr/local/share/slsh:/usr/local/lib/slsh:/usr/share/slsh:/usr/lib/slsh"; */
# define DEFAULT_CONF_PATH "/usr/local/etc:/usr/local/etc/slsh:/etc:/etc/slsh";
# define USER_SLSHRC ".slshrc"
#else
# define DEFAULT_CONF_PATH NULL
# define USER_SLSHRC "slsh.rc"
#endif

#ifdef __os2__
# ifdef __IBMC__
/* IBM VA3 doesn't declare S_IFMT */
#  define	S_IFMT	(S_IFDIR | S_IFCHR | S_IFREG)
# endif
#endif

#ifndef S_ISLNK
# ifdef S_IFLNK
#   define S_ISLNK(m) (((m) & S_IFMT) == S_IFLNK)
# else
#   define S_ISLNK(m) 0
# endif
#endif

#ifndef S_ISREG
# ifdef S_IFREG
#   define S_ISREG(m) (((m) & S_IFMT) == S_IFREG)
# else
#   define S_ISREG(m) 0
# endif
#endif

#ifndef S_ISDIR
# ifdef S_IFDIR
#   define S_ISDIR(m) (((m) & S_IFMT) == S_IFDIR)
# else
#   define S_ISDIR(m) 0
# endif
#endif

#ifndef S_ISCHR
# ifdef S_IFCHR
#   define S_ISCHR(m) (((m) & S_IFMT) == S_IFCHR)
# else
#   define S_ISCHR(m) 0
# endif
#endif

#ifndef S_ISBLK
# ifdef S_IFBLK
#   define S_ISBLK(m) (((m) & S_IFMT) == S_IFBLK)
# else
#   define S_ISBLK(m) 0
# endif
#endif

#ifndef S_ISFIFO
# ifdef S_IFIFO
#   define S_ISFIFO(m) (((m) & S_IFMT) == S_IFIFO)
# else
#   define S_ISFIFO(m) 0
# endif
#endif

#ifndef S_ISSOCK
# ifdef S_IFSOCK
#   define S_ISSOCK(m) (((m) & S_IFMT) == S_IFSOCK)
# else
#   define S_ISSOCK(m) 0
# endif
#endif

#ifndef S_IRUSR
# define S_IRUSR	0400
#endif
#ifndef S_IWUSR
# define S_IWUSR	0200
#endif
#ifndef S_IXUSR
# define S_IXUSR	0100
#endif
#ifndef S_IRGRP
# define S_IRGRP	0040
#endif
#ifndef S_IWGRP
# define S_IWGRP	0020
#endif
#ifndef S_IXGRP
# define S_IXGRP	0010
#endif
#ifndef S_IROTH
# define S_IROTH	0004
#endif
#ifndef S_IWOTH
# define S_IWOTH	0002
#endif
#ifndef S_IXOTH
# define S_IXOTH	0001
#endif
#ifndef S_ISUID
# define S_ISUID	04000
#endif
#ifndef S_ISGID
# define S_ISGID	02000
#endif
#ifndef S_ISVTX
# define S_ISVTX	01000
#endif

typedef struct _AtExit_Type
{
   SLang_Name_Type *nt;
   struct _AtExit_Type *next;
}
AtExit_Type;

static AtExit_Type *AtExit_Hooks;

static void at_exit (SLang_Ref_Type *ref)
{
   SLang_Name_Type *nt;
   AtExit_Type *a;

   if (NULL == (nt = SLang_get_fun_from_ref (ref)))
     return;

   a = (AtExit_Type *) SLmalloc (sizeof (AtExit_Type));
   if (a == NULL)
     return;

   a->nt = nt;
   a->next = AtExit_Hooks;
   AtExit_Hooks = a;
}

static void c_exit (int status)
{
   while (AtExit_Hooks != NULL)
     {
	AtExit_Type *next = AtExit_Hooks->next;
	if (SLang_get_error () == 0)
	  (void) SLexecute_function (AtExit_Hooks->nt);

	SLfree ((char *) AtExit_Hooks);
	AtExit_Hooks = next;
     }

   if (SLang_get_error ())
     SLang_restart (1);

   exit (status);
}

static void exit_intrin (void)
{
   int status;

   if (SLang_Num_Function_Args == 0)
     status = 0;
   else if (-1 == SLang_pop_int (&status))
     return;

   c_exit (status);
}

static void stat_mode_to_string (void)
{
   int mode, opts;
   char mode_string[12];

   opts = 0;
   if (SLang_Num_Function_Args == 2)
     {
	if (-1 == SLang_pop_integer (&opts))
	  return;
     }

   if (-1 == SLang_pop_integer (&mode))
     return;

   if (S_ISREG(mode)) mode_string[0] = '-';
   else if (S_ISDIR(mode)) mode_string[0] = 'd';
   else if (S_ISLNK(mode)) mode_string[0] = 'l';
   else if (S_ISCHR(mode)) mode_string[0] = 'c';
   else if (S_ISFIFO(mode)) mode_string[0] = 'p';
   else if (S_ISSOCK(mode)) mode_string[0] = 's';
   else if (S_ISBLK(mode)) mode_string[0] = 'b';

   if (mode & S_IRUSR) mode_string[1] = 'r'; else mode_string[1] = '-';
   if (mode & S_IWUSR) mode_string[2] = 'w'; else mode_string[2] = '-';
   if (mode & S_IXUSR) mode_string[3] = 'x'; else mode_string[3] = '-';
   if (mode & S_ISUID) mode_string[3] = 's';

#ifdef __WIN32__
   mode_string[4] = '-';
   mode_string[5] = '-';
   mode_string[6] = '-';

   if (opts & FILE_ATTRIBUTE_ARCHIVE) mode_string[7] = 'A'; else mode_string[7] = '-';
   if (opts & FILE_ATTRIBUTE_SYSTEM) mode_string[8] = 'S'; else mode_string[8] = '-';
   if (opts & FILE_ATTRIBUTE_HIDDEN) mode_string[9] = 'H'; else mode_string[9] = '-';
#else
   if (mode & S_IRGRP) mode_string[4] = 'r'; else mode_string[4] = '-';
   if (mode & S_IWGRP) mode_string[5] = 'w'; else mode_string[5] = '-';
   if (mode & S_IXGRP) mode_string[6] = 'x'; else mode_string[6] = '-';
   if (mode & S_ISGID) mode_string[6] = 'g';

   if (mode & S_IROTH) mode_string[7] = 'r'; else mode_string[7] = '-';
   if (mode & S_IWOTH) mode_string[8] = 'w'; else mode_string[8] = '-';
   if (mode & S_IXOTH) mode_string[9] = 'x'; else mode_string[9] = '-';
   if (mode & S_ISVTX) mode_string[9] = 't';
#endif

   mode_string[10] = 0;
   (void) SLang_push_string (mode_string);
}

#ifdef __WIN32__

/* Root/
 * 	bin/
 * 		slsh.exe
 * 	slsh/
 * 		slsh.rc
 * 		autoload.sl
 * 		( ... )
 */
static char *get_win32_root (void)
{
   static char base_path[MAX_PATH] = "";

   if (*base_path == 0)
     {
	if (GetModuleFileName(NULL, base_path, MAX_PATH-10) > 0)
	  {
	     /* drop file name */
	     char *p = strrchr(base_path, '\\');
	     if (p != NULL)
	       {
		  *p = 0;
		  /* drop also 'bin' */
		  p = strrchr(base_path, '\\');
		  if (p != NULL)
		    {
		       strcpy(p, "\\slsh\\");
		    }
	       }
	  }
     }

   return base_path;
}
#endif

static int Verbose_Loading;

static int try_to_load_file (char *path, char *file, char *ns)
{
   int status;

   if (path == NULL)
     path = ".";

   if (file != NULL)
     {
	file = SLpath_find_file_in_path (path, file);
	if (file == NULL)
	  return 0;
     }
   /* otherwise use stdin */

   status = SLns_load_file (file, ns);
   SLfree (file);
   if (status == 0)
     return 1;
   return -1;
}

static int load_startup_file (int is_interactive)
{
   char *dir;
   int status;

   dir = getenv (SLSH_CONF_DIR_ENV);
   if (dir == NULL)
     dir = getenv (SLSH_LIB_DIR_ENV);

   if (NULL == dir)
     {
#ifdef SLSH_CONF_DIR
	dir = SLSH_CONF_DIR;
	if (dir != NULL)
	  {
	     status = try_to_load_file (dir, SLSHRC_FILE, NULL);
	     if (status == -1)
	       return -1;
	     if (status == 1)
	       return 0;
	  }
#endif

#ifdef __WIN32__
	if (NULL == (dir = get_win32_root ()))
	  dir = DEFAULT_CONF_PATH;
#else
	dir = DEFAULT_CONF_PATH;
#endif
     }

   if (-1 == (status = try_to_load_file (dir, SLSHRC_FILE, NULL)))
     return -1;

   if ((status == 0)
       && (Verbose_Loading || is_interactive))
     {
	SLang_vmessage ("*** Installation Problem?  Unable to find the %s config file.",
			SLSHRC_FILE);
     }

   return 0;
}

#if 0
static int is_script (char *file)
{
   FILE *fp;
   char buf[3];
   int is;

   if (NULL == (fp = fopen (file, "r")))
     return 0;

   is = ((NULL != fgets (buf, sizeof(buf), fp))
	 && (buf[0] == '#') && (buf[1] == '!'));

   fclose (fp);
   return is;
}
#endif

static int setup_paths (void)
{
   char *libpath;

   if (NULL == (libpath = getenv (SLSH_PATH_ENV)))
     {
#ifdef SLSH_PATH
	libpath = SLSH_PATH;
#else
# ifdef __WIN32__
	libpath = get_win32_root ();
# endif
#endif
     }

   return SLpath_set_load_path (libpath);
}

static int set_verbose_loading (int *val)
{
   return SLang_load_file_verbose (*val);
}

/* Create the Table that S-Lang requires */
static SLang_Intrin_Fun_Type Intrinsics [] =
{
   MAKE_INTRINSIC_0("exit", exit_intrin, VOID_TYPE),
   MAKE_INTRINSIC_1("atexit", at_exit, VOID_TYPE, SLANG_REF_TYPE),
   MAKE_INTRINSIC_0("stat_mode_to_string", stat_mode_to_string, VOID_TYPE),
   MAKE_INTRINSIC_1("set_verbose_loading", set_verbose_loading, SLANG_INT_TYPE, SLANG_INT_TYPE),
   SLANG_END_INTRIN_FUN_TABLE
};

static void usage (void)
{
   char *libpath;
   fprintf (stderr, "\
Usage: slsh [OPTIONS] [-|file [args...]]\n\
 --help           Print this help\n\
 --version        Show slsh version information\n\
 -e string        Execute 'string' as S-Lang code\n\
 -g               Compile with debugging code, tracebacks, etc\n\
 -n               Don't load personal init file\n\
 --init file      Use this file instead of ~/%s\n",
	    USER_SLSHRC);
   fprintf (stderr, "\
 --no-readline    Do not use readline\n\
 -i               Force interactive input\n\
 -q, --quiet      Do not print startup messages\n\
 -t               Test mode.  If slsh_main exists, do not call it\n\
 -v               Show verbose loading messages\n\
 -Dname           Define \"name\" as a preprocessor symbol\n\
\n\
  Note: - and -i are mutually exclusive\n\
\n\
"
	    );
   libpath = SLpath_get_load_path ();
   fprintf (stderr, "Default search path: %s\n", (libpath == NULL) ? "" : libpath);
   SLang_free_slstring (libpath);

   exit (1);
}

static void output_version (void)
{
   fprintf (stdout, "slsh version %s; ", Slsh_Version);
   fprintf (stdout, "S-Lang version: %s\n", SLang_Version_String);
   if (SLANG_VERSION != SLang_Version)
     {
	fprintf (stdout, "\t** Note: This program was compiled against version %s.\n",
		 SLANG_VERSION_STRING);
     }
}

static int output_copyright (void)
{
   output_version ();
   fprintf (stdout, "Copyright (C) 2005-2011 John E. Davis <jed@jedsoft.org>\r\n");
   fprintf (stdout, "This is free software with ABSOLUTELY NO WARRANTY.\r\n");
   fprintf (stdout, "\n");

   return 0;
}

static void version (void)
{
   output_version ();
   exit (0);
}

int main (int argc, char **argv)
{
   char *file = NULL;
   char *init_file = USER_SLSHRC;
   char *init_file_dir;
   int exit_val;
   int is_interactive = 0;
   int use_readline = 1;
   int test_mode = 0;
   char *exec_string = NULL;
   int quiet = 0;

   (void) SLutf8_enable (-1);

   if ((-1 == SLang_init_all ())
       || (-1 == SLang_init_array_extra ())
       || (-1 == SLang_init_import ()) /* dynamic linking */
       || (-1 == SLadd_intrin_fun_table (Intrinsics, NULL))
       || (-1 == slsh_init_readline_intrinsics ()))
     {
	fprintf(stderr, "Unable to initialize S-Lang.\n");
	return 1;
     }

#ifdef SIGPIPE
   (void) SLsignal (SIGPIPE, SIG_IGN);
#endif

   /* FIXME for other systems */
#ifdef __WIN32__
   init_file_dir = getenv ("USERPROFILE");
#else
   init_file_dir = getenv ("HOME");
#endif

   if (-1 == setup_paths ())
     return -1;

#ifdef __WIN32__
   /* hack a directory for modules */
     {
	char *slshdir;
	char buffer[MAX_PATH] = "";

	slshdir = get_win32_root();
	if (slshdir != NULL)
	  strcpy(buffer, slshdir);
	strcat(buffer, "modules\\");
	SLang_set_module_load_path (buffer);
     }
#endif

   while (argc > 1)
     {
	char *arg = argv[1];

	if (0 == strcmp (arg, "--version"))
	  version ();

	if (0 == strcmp (arg, "--help"))
	  usage ();

	if (0 == strcmp (arg, "-i"))
	  {
	     argc--;
	     argv++;
	     is_interactive = 1;
	     continue;
	  }

	if ((0 == strcmp (arg, "-e"))
	    && (argc > 2))
	  {
	     argc -= 2;
	     argv += 2;
	     exec_string = *argv;
	     continue;
	  }

	if (0 == strcmp (arg, "-g"))
	  {
	     SLang_generate_debug_info (1);
	     SLang_Traceback = SL_TB_FULL;
	     argc--;
	     argv++;
	     continue;
	  }

	if (0 == strcmp (arg, "-n"))
	  {
	     init_file = NULL;
	     argc--;
	     argv++;
	     continue;
	  }

	if ((0 == strcmp (arg, "-q")) || (0 == strcmp (arg, "--quiet")))
	  {
	     quiet = 1;
	     argc--;
	     argv++;
	     continue;
	  }

	if (0 == strcmp (arg, "-t"))
	  {
	     test_mode = 1;
	     argc--;
	     argv++;
	     continue;
	  }

	if (0 == strcmp (arg, "-v"))
	  {
	     (void) SLang_load_file_verbose (3);
	     Verbose_Loading = 1;
	     argc--;
	     argv++;
	     continue;
	  }

	if (0 == strcmp (arg, "--no-readline"))
	  {
	     use_readline = 0;
	     argc--;
	     argv++;
	     continue;
	  }

	if ((0 == strcmp (arg, "--init"))
	    && (argc > 2))
	  {
	     init_file = argv[2];
	     init_file_dir = NULL;
	     argc -= 2;
	     argv += 2;
	     continue;
	  }

	if (0 == strncmp (arg, "-D", 2))
	  {
	     char *prep = arg + 2;
	     if (*prep != 0)
	       (void) SLdefine_for_ifdef (prep);
	     argc--;
	     argv++;
	  }

	break;
     }

   if (argc == 1)
     {
	if ((exec_string == NULL) && (is_interactive == 0))
	  is_interactive = (isatty (fileno(stdin)) && isatty (fileno(stdout)));
	file = NULL;
     }
   else
     {
	file = argv[1];
	if (0 == strcmp (file, "-"))
	  {
	     if (is_interactive)
	       usage ();
	     file = NULL;
	  }

	argc--;
	argv++;
     }

   if ((is_interactive == 0)
       && (SLang_Version < SLANG_VERSION))
     {
	fprintf (stderr, "***Warning: Executable compiled against S-Lang %s but linked to %s\n",
		 SLANG_VERSION_STRING, SLang_Version_String);
	fflush (stderr);
     }

   /* fprintf (stdout, "slsh: argv[0]=%s\n", argv[0]); */
   if (-1 == SLang_set_argc_argv (argc, argv))
     return 1;

   if (is_interactive)
     (void) SLdefine_for_ifdef ("__INTERACTIVE__");

   if (-1 == load_startup_file (is_interactive))
     return SLang_get_error ();

   /* Initializing the readline interface causes the .slrlinerc file
    * to be loaded.  It is put here after the startup files have been loaded.
    */
   if (-1 == slsh_use_readline (SLpath_basename (argv[0]), use_readline, is_interactive))
     return 1;

   if ((init_file != NULL)
       && (-1 == try_to_load_file (init_file_dir, init_file, NULL)))
     return SLang_get_error ();

   if ((file != NULL)
       || ((is_interactive == 0) && (exec_string == NULL)))
     {
	if (0 == try_to_load_file (NULL, file, NULL))
	  {
	     fprintf (stderr, "%s: file not found\n", file);
	     exit (1);
	  }
	if (test_mode == 0)
	  (void) SLang_run_hooks ("slsh_main", 0);
     }

   if (exec_string != NULL)
     {
	(void) SLang_load_string (exec_string);
     }

   if (is_interactive)
     {
	if (quiet == 0)
	  output_copyright ();
	if (SLang_Traceback != SL_TB_FULL)
	  SLang_Traceback = SL_TB_NONE;

	(void) slsh_interactive ();
     }

   exit_val = SLang_get_error ();
   c_exit (exit_val);
   return SLang_get_error ();
}
